core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/LogicalIndexExtractor.java

```diff
@@ -0,0 +1,378 @@
+            this.classifierSymbols = requireNonNull(classifierSymbols, "classifierSymbols is null");
+            this.matchNumberSymbols = requireNonNull(matchNumberSymbols, "matchNumberSymbols is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        protected Expression rewriteExpression(Expression node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            return treeRewriter.defaultRewrite(node, context);
+        }
+
+        @Override
+        public Expression rewriteLabelDereference(LabelDereference node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            Symbol reallocated = symbolAllocator.newSymbol(Symbol.from(node.getReference()));
+            layout.add(reallocated);
+            Set<IrLabel> labels = subsets.get(irLabel(node.getLabel()));
+            if (labels == null) {
+                labels = ImmutableSet.of(irLabel(node.getLabel()));
+            }
+            valuePointers.add(new ValuePointer(context.withLabels(labels).toLogicalIndexPointer(), Symbol.from(node.getReference())));
+            return reallocated.toSymbolReference();
+        }
+
+        @Override
+        public Expression rewriteSymbolReference(SymbolReference node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            // symbol reference with no label prefix is implicitly prefixed with a universal row pattern variable (matches every label)
+            // it is encoded as empty label set
+            Symbol reallocated = symbolAllocator.newSymbol(Symbol.from(node));
+            layout.add(reallocated);
+            valuePointers.add(new ValuePointer(context.withLabels(ImmutableSet.of()).toLogicalIndexPointer(), Symbol.from(node)));
+            return reallocated.toSymbolReference();
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            if (isPatternRecognitionFunction(node)) {
+                QualifiedName name = node.getName();
+                String functionName = name.getSuffix().toUpperCase(ENGLISH);
+                switch (functionName) {
+                    case "FIRST":
+                    case "LAST":
+                    case "PREV":
+                    case "NEXT":
+                        return rewritePatternNavigationFunction(node, context, treeRewriter);
+                    case "CLASSIFIER":
+                        return rewriteClassifierFunction(node, context);
+                    case "MATCH_NUMBER":
+                        return rewriteMatchNumberFunction();
+                }
+                throw new UnsupportedOperationException("unsupported pattern recognition function type: " + node.getName());
```

Comment:
We should keep track of this during analysis (i.e., a map of FunctionCall to type of pattern recognition function) so that we don't have to deal with identifiers, normalization, etc, in the planner. Sorry for not catching this before.

Reply:
This is not easy to do.

The pattern recognition expression goes through the following steps:
- analysis
- TranslationMap
- LogicalIndexExtractor.

We can record the function type by `NodeRef` in the Analysis. But then we lose the information in TranslationMap, where the function call is rewritten to a new Node (with rewritten args). The LogicalIndexExtractor gets the new node and has to resolve its type.

In order to carry the information to the LogicalIndexExtractor, we could:
- add a new field (e.g. `Optional<patternRecognitionFunctionType>`) in `FunctionCall` (which _possibly_ could be also useful when we support aggregations in pattern matching)
- inject the information about the new node back to the Analysis, which I consider unacceptable.

What do you think?

Reply:
Let's leave it as is.